File: packages/llm/src/score.ts
import { type Result, ok, err } from "@repo/shared";
import { complete } from "./client.js";

export interface ScoreInput {
  productName: string;
  summary: string;
  keyFeatures: string[];
  pros: string[];
  cons: string[];
  loadTimeMs: number;
}

export interface ScoreResult {
  overall: number;
  uxScore: number;
  performanceScore: number;
  featureScore: number;
  valueScore: number;
  reasoning: string;
  model: string;
}

function clamp(val: number, min: number, max: number): number {
  return Math.min(max, Math.max(min, val));
}

function parseJsonSafe(raw: string): Record<string, unknown> | null {
  try {
    return JSON.parse(raw);
  } catch {
    const match = raw.match(/\{[\s\S]*\}/);
    if (match) {
      try {
        return JSON.parse(match[0]);
      } catch {
        return null;
      }
    }
    return null;
  }
}

export async function scoreProduct(input: ScoreInput): Promise<Result<ScoreResult>> {
  const model = "claude-sonnet-4-5-20250929";
  const prompt = `You are a B2B software scoring engine. Score this product on a scale of 1-10 for each dimension.

Product: ${input.productName}
Summary: ${input.summary}
Key Features: ${input.keyFeatures.join(", ")}
Pros: ${input.pros.join(", ")}
Cons: ${input.cons.join(", ")}
Page Load Time: ${input.loadTimeMs}ms

Performance scoring guide:
- Under 1000ms load = 9-10
- 1000-2000ms = 7-8
- 2000-3000ms = 5-6
- Over 3000ms = 3-4

Respond in this exact JSON format (no markdown, just JSON):
{
  "overall": 7.5,
  "uxScore": 8.0,
  "performanceScore": 7.0,
  "featureScore": 8.0,
  "valueScore": 7.0,
  "reasoning": "Brief explanation of scores"
}`;

  const rawResult = await complete(prompt, { model });
  if (!rawResult.ok) {
    return rawResult;
  }

  const parsed = parseJsonSafe(rawResult.value);

  if (!parsed) {
    return err("Failed to parse LLM score response as JSON");
  }

  const uxScore = clamp(Number(parsed.uxScore) || 5, 1, 10);
  const performanceScore = clamp(Number(parsed.performanceScore) || 5, 1, 10);
  const featureScore = clamp(Number(parsed.featureScore) || 5, 1, 10);
  const valueScore = clamp(Number(parsed.valueScore) || 5, 1, 10);

  // Validate overall is roughly the average; if not, recalculate
  const avg = (uxScore + performanceScore + featureScore + valueScore) / 4;
  let overall = clamp(Number(parsed.overall) || avg, 1, 10);
  if (Math.abs(overall - avg) > 1.5) {
    overall = Math.round(avg * 10) / 10;
  }

  return ok({
    overall,
    uxScore,
    performanceScore,
    featureScore,
    valueScore,
    reasoning: String(parsed.reasoning ?? ""),
    model,
  });
}

File: apps/go-api/internal/llm/score.go
package llm

import (
	"context"
	"encoding/json"
	"fmt"
	"math"
	"strings"
)

// ScoreInput contains the data needed to generate product scores.
type ScoreInput struct {
	ProductName string
	Summary     string
	KeyFeatures []string
	Pros        []string
	Cons        []string
	LoadTimeMs  int
}

// ScoreResult holds the structured scoring output.
type ScoreResult struct {
	Overall          float64 `json:"overall"`
	UXScore          float64 `json:"uxScore"`
	PerformanceScore float64 `json:"performanceScore"`
	FeatureScore     float64 `json:"featureScore"`
	ValueScore       float64 `json:"valueScore"`
	Reasoning        string  `json:"reasoning"`
	Model            string  `json:"model"`
}

// Score generates numerical scores for a product based on its summary data.
// Validates that the overall score is within 1.5 of the average of sub-scores.
func (c *Client) Score(ctx context.Context, input ScoreInput) (ScoreResult, error) {
	prompt := fmt.Sprintf(`You are a professional B2B software reviewer scoring a product. Rate each dimension from 1.0 to 10.0.

Product: %s
Summary: %s
Key Features: %s
Pros: %s
Cons: %s
Page Load Time: %dms

Respond with ONLY valid JSON in this exact format:
{
  "overall": 7.5,
  "uxScore": 7.0,
  "performanceScore": 8.0,
  "featureScore": 7.5,
  "valueScore": 7.0,
  "reasoning": "Brief explanation of the scores"
}`,
		input.ProductName,
		input.Summary,
		strings.Join(input.KeyFeatures, ", "),
		strings.Join(input.Pros, ", "),
		strings.Join(input.Cons, ", "),
		input.LoadTimeMs,
	)

	raw, err := WithRetry(ctx, func() (string, error) {
		return c.Complete(ctx, prompt, WithSystem("You are a B2B SaaS product scorer. Always respond with valid JSON only."))
	}, DefaultRetryOptions())
	if err != nil {
		return ScoreResult{}, fmt.Errorf("score completion: %w", err)
	}

	var result ScoreResult
	if err := json.Unmarshal([]byte(raw), &result); err != nil {
		return ScoreResult{}, fmt.Errorf("parsing score JSON: %w (raw: %.200s)", err, raw)
	}

	// Clamp all scores to [1, 10].
	result.Overall = clamp(result.Overall, 1, 10)
	result.UXScore = clamp(result.UXScore, 1, 10)
	result.PerformanceScore = clamp(result.PerformanceScore, 1, 10)
	result.FeatureScore = clamp(result.FeatureScore, 1, 10)
	result.ValueScore = clamp(result.ValueScore, 1, 10)

	// Validate overall vs average of sub-scores (tolerance ±1.5).
	avg := (result.UXScore + result.PerformanceScore + result.FeatureScore + result.ValueScore) / 4.0
	if math.Abs(result.Overall-avg) > 1.5 {
		result.Overall = math.Round(avg*10) / 10
	}

	result.Model = "claude-sonnet-4-20250514"
	return result, nil
}

func clamp(v, min, max float64) float64 {
	if v < min {
		return min
	}
	if v > max {
		return max
	}
	return v
}
